File: hash-table-medium/36.有效的数独.py
#
# @lc app=leetcode.cn id=36 lang=python3
#
# [36] 有效的数独
#

# @lc code=start
class Solution:
    # 507/507 cases passed (36 ms)
    # Your runtime beats 99.38 % of python3 submissions
    # Your memory usage beats 65.13 % of python3 submissions (14.8 MB)
    def isValidSudoku(self, board: list[list[str]]) -> bool:
        # Init data
        rows = [dict() for i in range(9)]
        cols = [dict() for i in range(9)]
        boxes = [dict() for i in range(9)]
        for i in range(9):
            for j in range(9):
                num = board[i][j]
                if num != ".":
                    box_index = (i // 3) * 3 + (j // 3)
                    # Fill rows, cols, boxes
                    rows[i][num] = rows[i].get(num, 0) + 1
                    cols[j][num] = cols[j].get(num, 0) + 1
                    boxes[box_index][num] = boxes[box_index].get(num, 0) + 1
                    # Check duplicate
                    if (
                        rows[i].get(num) > 1
                        or cols[j].get(num) > 1
                        or boxes[box_index].get(num) > 1
                    ):
                        return False
        return True


# @lc code=end


class Solution_1:
    def isValidSudoku(self, board: list[list[str]]) -> bool:
        matrix_line = [set() for i in range(9)]
        matrix_column = [set() for i in range(9)]
        matrix_area = [set() for i in range(9)]
        for i in range(9):
            for j in range(9):
                item = board[i][j]
                pos = (i // 3) * 3 + j // 3
                if item != ".":
                    if (
                        item not in matrix_line[i]
                        and item not in matrix_column[j]
                        and item not in matrix_area[pos]
                    ):
                        matrix_line[i].add(item)
                        matrix_column[j].add(item)
                        matrix_area[pos].add(item)
                    else:
                        return False
        return True

File: hash-table-medium/36.有效的数独.java
/*
 * @lc app=leetcode.cn id=36 lang=java
 *
 * [36] 有效的数独
 */

// @lc code=start
class Solution {
    // 507/507 cases passed (2 ms)
    // Your runtime beats 95.29 % of java submissions
    // Your memory usage beats 85.63 % of java submissions (38.3 MB)
    public boolean isValidSudoku(char[][] board) {
        boolean[][] rows = new boolean[9][9];
        boolean[][] cols = new boolean[9][9];
        boolean[][] boxes = new boolean[9][9];
        
        for (int i = 0; i <9; i++) {
            for (int j = 0; j <9; j++) {
                if (board[i][j] == '.') {
                    continue;
                }
                int box_index = i / 3 * 3 + j / 3;
                int num = board[i][j] - '1';
                if (rows[i][num] || cols[j][num] || boxes[box_index][num]) {
                    return false;
                } else {
                    rows[i][num] = true;
                    cols[j][num] = true;
                    boxes[box_index][num] = true;
                }
            }
        }
        return true;
    }
}
// @lc code=end

class Solution {
    public boolean isValidSudoku(char[][] board) {
        // 记录某行，某位数字是否已经被摆放
        boolean[][] row = new boolean[9][9];
        // 记录某列，某位数字是否已经被摆放
        boolean[][] col = new boolean[9][9];
        // 记录某 3x3 宫格内，某位数字是否已经被摆放
        boolean[][] block = new boolean[9][9];

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] != '.') {
                    int num = board[i][j] - '1';
                    int blockIndex = i / 3 * 3 + j / 3;
                    if (row[i][num] || col[j][num] || block[blockIndex][num]) {
                        return false;
                    } else {
                        row[i][num] = true;
                        col[j][num] = true;
                        block[blockIndex][num] = true;
                    }
                }
            }
        }
        return true;
    }
}